File: ptolemy/domains/od/kernel/Event.java
/* An aggregation consisting of a Token, a time stamp and destination Receiver.

 Copyright (c) 1997-1998 The Regents of the University of California.
 All rights reserved.
 Permission is hereby granted, without written agreement and without
 license or royalty fees, to use, copy, modify, and distribute this
 software and its documentation for any purpose, provided that the above
 copyright notice and the following two paragraphs appear in all copies
 of this software.

 IN NO EVENT SHALL THE UNIVERSITY OF CALIFORNIA BE LIABLE TO ANY PARTY
 FOR DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
 ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF
 THE UNIVERSITY OF CALIFORNIA HAS BEEN ADVISED OF THE POSSIBILITY OF
 SUCH DAMAGE.

 THE UNIVERSITY OF CALIFORNIA SPECIFICALLY DISCLAIMS ANY WARRANTIES,
 INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE SOFTWARE
 PROVIDED HEREUNDER IS ON AN "AS IS" BASIS, AND THE UNIVERSITY OF
 CALIFORNIA HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES,
 ENHANCEMENTS, OR MODIFICATIONS.

                                        PT_COPYRIGHT_VERSION_2
                                        COPYRIGHTENDKEY

@ProposedRating Red (davisj@eecs.berkeley.edu)
*/

package ptolemy.domains.od.kernel;

import ptolemy.actor.*;
import ptolemy.data.*;

//////////////////////////////////////////////////////////////////////////
//// Event
/**
An Event is an aggregation consisting of a Token, a time stamp and 
destination Receiver. Both the token and destination receiver are 
allowed to have null values. This is particularly useful in situations
where the specification of the destination receiver may be considered 
redundant.

@author John S. Davis II
@version @(#)Event.java	1.3	11/02/98
@see ptolemy.actors.Token
*/

public class Event {

    /** Construct an Event with a token and time stamp.
     */
    public Event(Token token, double time) {
        _token = token;
        _timeStamp = time;
    }

    /** Construct an Event with a token, a time stamp and a 
     *  destination receiver.
     */
    public Event(Token token, double time, Receiver receiver) {
        this(token, time);
        _receiver = receiver;
    }


    ///////////////////////////////////////////////////////////////////
    ////                         public methods                    ////

    /** Return the destination receiver of this event.
     *  @return Receiver The destination receiver of this event.
     */
    public Receiver getReceiver() {
        return _receiver;
    }

    /** Return the time stamp of this event.
     *  @return double The time stamp of this event.
     */
    public double getTime() {
        return _timeStamp;
    }

    /** Return the token of this event.
     *  @return Token The token of this event.
     */
    public Token getToken() {
        return _token;
    }

    ///////////////////////////////////////////////////////////////////
    ////                         private variables                 ////

    double _timeStamp = 0.0;
    Token _token = null;
    Receiver _receiver = null;
}

File: ptII/ptolemy/apps/p1000/c/demo/scheduler/event.c
/* Event that can be sent and received via connections between ports.

 Copyright (c) 1997-2006 The Regents of the University of California.
 All rights reserved.
 Permission is hereby granted, without written agreement and without
 license or royalty fees, to use, copy, modify, and distribute this
 software and its documentation for any purpose, provided that the above
 copyright notice and the following two paragraphs appear in all copies
 of this software.

 IN NO EVENT SHALL THE UNIVERSITY OF CALIFORNIA BE LIABLE TO ANY PARTY
 FOR DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES
 ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF
 THE UNIVERSITY OF CALIFORNIA HAS BEEN ADVISED OF THE POSSIBILITY OF
 SUCH DAMAGE.

 THE UNIVERSITY OF CALIFORNIA SPECIFICALLY DISCLAIMS ANY WARRANTIES,
 INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE SOFTWARE
 PROVIDED HEREUNDER IS ON AN "AS IS" BASIS, AND THE UNIVERSITY OF
 CALIFORNIA HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES,
 ENHANCEMENTS, OR MODIFICATIONS.

 PT_COPYRIGHT_VERSION_2
 COPYRIGHTENDKEY

 @author Thomas Huining Feng

 */

#include <stdlib.h>
#include <string.h>
#include "event.h"

/**
 * Duplicate an event in the heap, and return the duplicated event. If the given
 * event contains a token, memory will be allocated from the heap to store a
 * copy of that token. The user need not explicitly free the token, though
 * he/she do need to free the event, because the event and the token are stored
 * in a continuous trunk of memory.
 * 
 * @param event Reference to the event to be duplicated.
 * @return Reference to the duplicated event in the heap.
 */
Event* Event_duplicate(const Event* event) {
	Event* heap_event;
	void* heap_token;
	size_t token_size;
	
	token_size = event->token == NULL ? 0
			: ((GeneralType*) event->token)->typeData->size;
	heap_event = (Event*) malloc(sizeof(Event) + token_size);

	// Forward compatible with future event definitions, which may contain more
	// fields.
	*heap_event = *event;
	
	if (event->token != NULL) {
		// Copy the token in to heap so that the original event's token can be
		// freed immediately.
		heap_token = ((void*) heap_event) + sizeof(Event);
		memcpy(heap_token, event->token, token_size);
		heap_event->token = (Token*) heap_token;
	}
	
	return heap_event;
}
